Remove all installers and install logic

mjlbach opened the pull request:
Closes #334 #26 #92 #289 #359 #361 #421 #467

mjlbach commented:
@teto This is going to take fairly exhaustive testing, and we'll probably have to update the documentation. It's mostly ready, and I think I removed all dead documentation. To be honest, I don't know how to install a lot of these language servers the "normal" way, since all of mine are packaged using nix.

pretentious7 commented:
:clap: :clap: :clap: 

I've been using basically that with typescript for a couple days, so can confirm that works.

teto commented on lua/lspconfig/jsonls.lua:
```diff
@@ -35,7 +18,7 @@ https://github.com/vscode-langservers/vscode-json-languageserver
 vscode-json-languageserver, a language server for JSON and JSON schema
 
-`vscode-json-languageserver` can be installed via `:LspInstall jsonls` or by yourself with `npm`:
+`vscode-json-languageserver` can be install via `npm`:
```
installed ?

teto commented on lua/lspconfig/metals.lua:
```diff
@@ -94,13 +29,11 @@ let g:metals_server_version = '0.8.4+106-5f2b9350-SNAPSHOT'
 Scala language server with rich IDE features.
-`metals` can be installed via `:LspInstall metals`.
+
+See project page for installation instructions
```
maybe reference that page ?

mjlbach replied:
Done, and I added instruction. It's absurdly difficult to find documentation for how to manually install their language server, I had to look under the emacs instructions for a link, I copied the installation steps just to be safe).

teto commented:
thanks for tackling this, it needs to be done before the 0.5 release so that users dont start relying on these primitives.
Once this is merged, we need to reference it in https://github.com/neovim/neovim/wiki/Following-HEAD .

mjlbach commented:
If there are any remaining dicey documentation issues let me know. I think we should also fix whatever is going on with the print warning that cmd is undefined before merging (and hopefully more people chime in that the instructions are great/bad/insufficient. 

This will also break things for people, as cmd will need to be defined. So I imagine an influx of issues. I don't think there's a good intermediate measure.

teto commented:
I think we should default `cmd` to the name of the binary and let the user put it in PATH. The good thing about this is that instead of a (lua) error (?), we can advise to run `:checkhealth nvim_lsp` and it will tell the user that the binary is not in PATH ? Ideally we need a much more helpful checkhealth for lsp https://github.com/neovim/neovim/pull/13161

mjlbach commented:
> I think we should default cmd to the name of the binary and let the user put it in PATH.

I initially tried doing this, but many "commands" are simply the path to some random output of an npm build process. My strategy was mostly to keep the names of commands that point to actual binaries one would reasonably put on their path and remove those that were some random build output file. Also there's a slight issue in that windows binaries often will have a different extension. Right now without the error handling I put in, you'll get a nasty red error message if the language client fails to start the server.

ckipp01 commented on lua/lspconfig/metals.lua:
````diff
@@ -79,28 +17,36 @@ configs[server_name] = {
 Scala language server with rich IDE features.
-`metals` can be installed via `:LspInstall metals`.
+
+See Emacs instructions in the metals documentation:
+
+https://scalameta.org/metals/docs/editors/emacs.html
+
+The installation instructions are replicated here.
+Note that you can place the binary anywhere on your path.
+
+```bash
+# Make sure to use coursier v1.1.0-M9 or newer.
+curl -L -o coursier https://git.io/coursier-cli
+chmod +x coursier
+./coursier bootstrap \
+  --java-opt -Xss4m \
+  --java-opt -Xms100m \
+  --java-opt -Dmetals.client=emacs \
+  org.scalameta:metals_2.12:0.9.8 \
+  -r bintray:scalacenter/releases \
+  -r sonatype:snapshots \
+  -o /usr/local/bin/metals -f
+```
````
I would change this as bootstrapping metals with `-Dmetals.client=emacs` sets some defaults that will cause Metals to not really work as expected. There actually is a very small section on using other vim LSP clients on the vim page here https://scalameta.org/metals/docs/editors/vim.html#using-an-alternative-lsp-client. However, that will be updated soon with nvim LSP specific stuff as well, since we're actually putting effort into having a nvim LSP focused extension for Metals users: https://github.com/scalameta/nvim-metals. This is mainly due to a  lot of defaults not working unless some extra things are offered, and we make heavy use of LSP extensions. Also an alternative to what you have here can I suggest the following?

```suggestion
See full instructions in the Metals documentation:

https://scalameta.org/metals/docs/editors/vim.html#using-an-alternative-lsp-client

Note: that if you're using [nvim-metals](https://github.com/scalameta/nvim-metals), that plugin fully handles the setup and installation of Metals, and you shouldn't set up Metals both with it and this plugin.

To install Metals, make sure to have [coursier](https://get-coursier.io/docs/cli-installation) installed, and once you do you can install the latest Metals with `cs install metals`. You can also manually bootstrap Metals with the following command.

```bash
cs bootstrap \
  --java-opt -Xss4m \
  --java-opt -Xms100m \
  org.scalameta:metals_2.12:<enter-version-here> \
  -r bintray:scalacenter/releases \
  -r sonatype:snapshots \
  -o /usr/local/bin/metals -f
```
```

mjlbach commented:
@ckipp01 Thanks for the useful feedback! I've merged your changes. I don't use metals, and had a bit of a hard time finding the page you referenced, so I'm glad that is being worked on. I'm happy to nudge people to the nvim-metals plugin in the documentation; I don't think we can ever match the deep integration that server specific plugins have.

Let me know if there are any specific features you want that would make the plugin easier. I've recently filed several PRs with core improving 3.16 compliance.

mjlbach commented:
@teto Unless I'm mistaken, I think the only bin_name we're missing now is elixir

mjlbach commented:
@mortepau Do you have suggestions for improving the documentation regarding elixir install and default cmd without LspInstall?

saadparwaiz1 commented:
@mjlbach I'm not sure if this is helpful but I have been using the elixir language server with emacs's lsp-mode and the installation instruction are as simple picking up the zip file from github release and using language_server.sh on unix and language_server.bat on windows. These can be found [here](https://github.com/elixir-lsp/elixir-ls#detailed-installation-instructions)

mjlbach commented:
@teto I think this is good for another look. I added the missing elixir/sumneko lua documentation, and added some documentation that had always been missing for OmniSharp.

teto commented:
Is it possible to fix the CONFIG.md conflict. I will let @h-michael merge this when he feels like it, he is the most knowledgeable about this repo.

mjlbach commented:
Fixed

teto commented on lua/lspconfig.lua:
```diff
@@ -4,71 +4,10 @@ local M = {
--- script in scriptnames to be executed is lspconfig.
-function M._root._setup()
-  M._root.commands = {
-    LspInstall = {
```
neovim 0.5 hasn't been released yet so in theory we could just get rid of this but aving the commands print a message like `print("deprecated, see https://github.com/neovim/neovim/wiki/Following-HEAD")  would avoid some pain for everyone (users and contributors on gitter). And we would remove it just before 0.5.

mjlbach replied:
Added in a separate commit to make it easy to remove 🔪

teto commented on lua/lspconfig/als.lua:
```diff
@@ -7,78 +7,19 @@ if vim.fn.has('win32') == 1 then
 https://github.com/AdaCore/ada_language_server
 
-Ada language server. Use `LspInstall als` to install it.
+See project page for installation instructions.
```
https://github.com/AdaCore/ada_language_server#Install

teto reviewed (approved):
I've grepped the source for LspInstall and there is one mention left in health.lua

mjlbach commented:
I noticed that one, but I'm not sure what it means...

teto commented:
git blame shows it was due to https://github.com/neovim/nvim-lspconfig/issues/248 . Seems like make_config  is not available prior to te setup call ? maybe keep the `if`  and just remove the comment about LspInstall for now. I think we can merge this today,

mjlbach commented:
Done, I reworded it to reflect what I (think?) is the purpose.

teto commented:
cool, can you mention the change on https://github.com/neovim/neovim/wiki/Following-HEAD ? feel free to merge when it's done

mjlbach commented:
Done, and merging.

Borwe commented:
Man I miss the auto LspInstall... was so easy.

teto commented:
I've seen https://github.com/anott03/nvim-lspinstall popup in my github notifs. Dont know how good it is but may be a solution.
